backend/django/create-model-and-api.mo.md

```diff
@@ -0,0 +1,219 @@
+```python
+from django.apps import AppConfig
+from django.contrib.admin import site
+from django.db.models.signals import post_migrate
+
+
+class UsersConfig(AppConfig):
+    name = 'intramuros.users'
+    verbose_name = "Users"
+
+    def ready(self):
+        post_migrate.connect(add_group_permissions, sender=self)
+
+
+def add_model_permissions(group, model, ContentType, Permission):
+    content_type = ContentType.objects.get_for_model(model)
+    permissions = Permission.objects.filter(content_type=content_type)
+    for permission in permissions:
+        group.permissions.add(permission)
+
+
+def add_group_permissions(sender, using, apps, **kwargs):
+    Group = apps.get_model("auth", "Group")
+    Permission = apps.get_model("auth", "Permission")
+    ContentType = apps.get_model("contenttypes", "ContentType")
+    News = apps.get_model("publications", "News")   #Add this line
+
+    if Group.objects.using(using).filter(name='Mayor Admin').exists():
+        group = Group.objects.get(name='Mayor Admin')
+    else:
+        group = Group.objects.using(using).create(name='Mayor Admin')
+    add_model_permissions(group, News, ContentType, Permission) #Add this second line
+```
+
+- Finally, make your migrations and migrate.
+
+
+### Add the News model to the admin *(~2 min)*
+
+- We now need to add the model to the admin back-office.
+
+ourproject/publications/admin.py
+```python
+from django.contrib import admin
+from .models import News
+
+
+@admin.register(News)
+class MyNewsAdmin(admin.ModelAdmin):
+    list_display = (
+        'title',
+        'city'
+    )
+```
+
+If you don't override `list_display` by writing `pass` instead of `list_display`, the default column title in the admin will be the __str__ returned in the model.
+list_display allows you to add several columns to the admin interface.
+
+
+### Serialize the News you get from the database *(~2 min)*
+
+- Serializers define the API representation.
+
+
+ourproject/publications/serializers.py
+```python
+from .models import News
+from rest_framework import serializers
+
+
+class NewsSerializer(serializers.HyperlinkedModelSerializer):
+    id = serializers.IntegerField(source='pk')
+    city = serializers.StringRelatedField(many=False, source="city.pk")   #We need this to get the primary key of the city that is attached to this news
+
+    class Meta:
+        model = News
+        fields = ('title', 'description', 'created_at', 'updated_at', 'city', 'id')    #All the fields you wish to get
+```
+
+
+### Create a News ViewSet *(~2 min)*
+
+
+```python
+from .models import News
+from rest_framework import viewsets
+from .serializers import NewsSerializer
+
+#This is an example of filtering the pieces of news with the city-id like this: /?city-id=5
+class FilterByCity(object):   
+    def get_queryset(self):   #You need to override the default get_queryset method
+        queryset = super().get_queryset()
+        city_id = self.request.query_params.get('city-id')
```

Comment:
This code filters ? it's not clear in the explanation

Reply:
# This tries to get the request param `city-id`. If you make the request GET localhost/news?city-id=4, this piece of code returns 4. If there is no param, it returns `null`